technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+# K8ssandra Operator Design
+
+Author: @jsanda
+
+## Overview
+
+GH issue: https://github.com/k8ssandra/k8ssandra/issues/683
+
+This is a design document for an operator for k8ssandra. For background see 
+https://github.com/k8ssandra/k8ssandra/issues/485.
+
+Much of the behavior that is implemented in the Helm charts will be reimplemented in various controllers. The role of the Helm charts changes to focus primarily on installation and configuration of operators.
+
+While we will continue to use and support Helm, it will not be required for installing k8ssandra from a purely technical perspective. Users can install manifests with kubectl apply for example. We can also provide support for kustomize.
+
+## CustomResourceDefinitions
+
+We will utilize a number of CRDs, some of which are developed and maintained by other projects.
```

Comment:
I guess instead of shipping with pre-generated yamls from other projects, we should look at including only the api modules through go dependencies and do go generate to get the yaml files. That means we should probably have proper annotations in all those apis (such as //+kubebuilder:rbac styles)

Reply:
Of course, if we include reaper & medusa operator to this single operator, then that's automated process - but for cass-operator resources we can do the same. After that, only thing required would be to update the go.mod

Reply:
If - and only if - we don't do kustomize templates in cass-operator that modify the generated CRD. Which we actually might want to do for some kube-x- properties, so keep the generated yaml for cass-operator.

Reply:
I think we should provide kustomize support for cass-operator as well. As for some of the kube-x properties, we can apply patches with kustomize to handle them. We already do this with medusa-operator.